### PR TITLE
Support opening anchor channels and test end-to-end unilateral close

### PR DESCRIPTION
**Pull request by wpaulino**
This PR introduces support for the anchors feature bit and opening/accepting anchor channels. It also includes an end-to-end test of the unilateral channel close flow, consuming the recently added anchor events and attaching fees to transactions when necessary.

Depends on https://github.com/lightningdevkit/rust-lightning/pull/1825.

### CODE REVIEW

**Comment by ariard** [comment]
Note, there was a security issue affecting LND/Core Lightning anchor output support few months ago: https://lists.linuxfoundation.org/pipermail/lightning-dev/2022-April/003561.html. Briefly, anchor output relaxes the sighash flags for second-stage HTLC transactions from `SIGHASH_ALL` to `SIGHASH_SINGLE | SIGHASH_ANYONECANPAY`, allowing your counterparty to combine them in a single-batch. If your implementation was doing assumptions on the inputs/outputs size or order to parse those second-stage HTLC transactions in case of revoked state, a malicious party could extract funds.

With the current state of our parsing logic in `check_spend_counterparty_htlc`, I think we're encumbering the same exact security issue due to the following checks (1231907, `channelmonitors`, L2644) and we should adapt our logic in consequence:

```
		if tx.input.len() != 1 || tx.output.len() != 1 || tx.input[0].witness.len() != 5 {
```

**Comment by ariard** [comment]
More test coverage requested from #1825: https://github.com/lightningdevkit/rust-lightning/pull/1825#discussion_r1040277311

File: lightning/src/ln/features.rs [hunk 1]
```diff
@@ -52,6 +52,8 @@
 //!     (see [BOLT-2](https://github.com/lightning/bolts/blob/master/02-peer-protocol.md) for more information).
 //! - `Keysend` - send funds to a node without an invoice
 //!     (see the [`Keysend` feature assignment proposal](https://github.com/lightning/bolts/issues/605#issuecomment-606679798) for more information).
+//! - `AnchorsZeroFeeHtlcTx` - requires/supports that commitment transactions include anchor outputs and HTLC transactions have zero fee
```
Comment (ariard) [inline lightning/src/ln/features.rs:55]: Can say "HTLC transactions are pre-signed with zero fee" The transactions they have fees, they're just added a posteriori

File: lightning/src/ln/features.rs [hunk 2]
```diff
@@ -346,6 +348,9 @@ mod sealed {
 	define_feature!(19, Wumbo, [InitContext, NodeContext],
 		"Feature flags for `option_support_large_channel` (aka wumbo channels).", set_wumbo_optional, set_wumbo_required,
 		supports_wumbo, requires_wumbo);
+	define_feature!(23, AnchorsZeroFeeHtlcTx, [InitContext, NodeContext, ChannelTypeContext],
```
Comment (ariard) [inline lightning/src/ln/features.rs:351]: According to BOLT9, `option_anchors_zero_fee_htlc_tx` is IN, so I don't know if we should be only `InitContext` / `NodeContext` here? The spec feature bit context nomenclature is obscure, without that much ground so I suppose it could also make sense to amend the spec here.

File: lightning/src/util/config.rs [hunk 3]
```diff
@@ -149,7 +148,22 @@ pub struct ChannelHandshakeConfig {
...
+	/// If set, we attempt to negotiate the `anchors_zero_fee_htlc_tx`option for outbound channels.
+	///
+	/// If this option is set, channels may be created that will not be readable by LDK versions
+	/// prior to 0.0.110, causing [`ChannelManager`]'s read method to return a
+	/// [`DecodeError::InvalidValue`].
+	///
+	/// Note that setting this to true does *not* prevent us from opening channels with
+	/// counterparties that do not support the `anchors_zero_fee_htlc_tx` option; we will simply
+	/// fall back to a `static_remote_key` channel.
+	///
```
Comment (ariard) [inline lightning/src/util/config.rs:161]: I think we could add an implementation disclaimer we do not aim to support simple `option_anchor_outputs` and we can add historical context on the rational: https://lists.linuxfoundation.org/pipermail/lightning-dev/2020-September/002796.html

File: lightning/src/ln/channel.rs [hunk 4]
```diff
@@ -2081,7 +2093,12 @@ impl<Signer: Sign> Channel<Signer> {
 		} else if their_features.supports_channel_type() {
 			// Assume they've accepted the channel type as they said they understand it.
 		} else {
-			self.channel_type = ChannelTypeFeatures::from_counterparty_init(&their_features)
+			self.channel_type = ChannelTypeFeatures::from_counterparty_init(&their_features);
+			self.channel_transaction_parameters.opt_anchors = if self.channel_type.supports_anchors_zero_fee_htlc_tx() {
+					Some(())
+			} else {
+					None
```
Comment (ariard) [inline lightning/src/ln/channel.rs:2100]: I think here we can have a situation where our counterparty assumes we have another channel type like static_remote_key or simple anchor, and that could lead to a signature wrong on the first channel commitment transaction exchanged, am i correct ? In case of feature negotiation failure we could also fail the channel opening and give back an error message, I guess.

File: lightning/src/ln/channel.rs [hunk 5]
```diff
@@ -1136,9 +1150,7 @@ impl<Signer: Sign> Channel<Signer> {
 		} else {
 			ChannelTypeFeatures::from_counterparty_init(&their_features)
 		};
-		if !channel_type.supports_static_remote_key() {
-			return Err(ChannelError::Close("Channel Type was not understood - we require static remote key".to_owned()));
-		}
+		let opt_anchors = channel_type.supports_anchors_zero_fee_htlc_tx();
```
Comment (ariard) [inline lightning/src/ln/channel.rs:1153]: The test case which could be valuable would be actually someone feeding opening/accepting messages with `option_anchor_outputs` and check we either drop the funding flow or overrule to the safe `option_anchor_outputs`. Being robust against channel type safe downgrade attempts.

**Review by ariard — COMMENTED** [review]
Left few reviews comments.

> There are definitely several more cases we should be testing, but in the spirit of keeping PRs relatively small, this PR will just focus on that edge case and the happy case.

We can even move the aggregated revoked HTLC test in its own PR, bundled with few others from #1825. I think it's already good if we focus just on all the cases of features bit interpretation in both directions here, including unsafe `option_anchor_outputs` attempts.

File: lightning/src/ln/features.rs (hunk 2 above)
Comment (ariard) [reply lightning/src/ln/features.rs:355]: Yes, effectively the `channel_type` field is described in BOLT2 without back reference in BOLT9 -- For the confusion of everyone.

File: lightning/src/ln/channel.rs [hunk 6]
```diff
@@ -898,7 +909,17 @@ impl<Signer: Sign> Channel<Signer> {
...
+		// features one by one until we've either arrived at our default or the counterparty has
+		// accepted one.
+		if self.channel_type.supports_anchors_zero_fee_htlc_tx() {
+			self.channel_type.clear_anchors_zero_fee_htlc_tx();
+			self.channel_transaction_parameters.opt_anchors = None;
+		} else if self.channel_type.supports_scid_privacy() {
+			self.channel_type.clear_scid_privacy();
+		} else {
+			self.channel_type = ChannelTypeFeatures::only_static_remote_key();
+		}
```
Comment (ariard) [inline lightning/src/ln/channel.rs:922]: Should we assert `ChannelTransactionParameters.opt_non_zero_fee_anchors` is `None` here for sanity ?

File: lightning/src/util/config.rs [hunk 7]
```diff
@@ -149,7 +148,28 @@ pub struct ChannelHandshakeConfig {
 	///                as 1000 sats instead, which is a safe implementation-specific lower bound.
 	/// Maximum value: 1,000,000, any values larger than 1 Million will be treated as 1 Million (or 100%)
 	///                instead, although channel negotiations will fail in that case.
-	pub their_channel_reserve_proportional_millionths: u32
+	pub their_channel_reserve_proportional_millionths: u32,
+	#[cfg(anchors)]
+	/// If set, we attempt to negotiate the `anchors_zero_fee_htlc_tx`option for outbound channels.
```
Comment (ariard) [inline lightning/src/util/config.rs:153]: I think here we can document LDK behavior in case of inbound channels not supporting `anchors_zero_fee_htlc_tx`, for now we don't provide an option to reject those ones and this feature likely coming in the future.

File: lightning/src/ln/channel.rs [hunk 8]
```diff
@@ -1134,31 +1163,28 @@ impl<Signer: Sign> Channel<Signer> {
...
-				}
-
-				if channel_type.requires_scid_privacy() && announced_channel {
-					return Err(ChannelError::Close("SCID Alias/Privacy Channel Type cannot be set on a public channel".to_owned()));
-				}
+			// Make sure we support all of the features behind the channel type.
+			if !channel_type.is_subset(our_supported_features) {
+				return Err(ChannelError::Close("Channel Type contains unsupported features".to_owned()));
+			}
+			if channel_type.requires_scid_privacy() && announced_channel {
```
Comment (arik-so) [inline lightning/src/ln/channel.rs:1176]: Should we have a method that finds other possible feature flag inconsistencies to surface illegal combinations?

File: lightning/src/ln/features.rs [hunk 9]
```diff
@@ -665,6 +672,24 @@ impl<T: sealed::Context> Features<T> {
 			(byte & unknown_features) != 0
 		})
 	}
+
+	// Returns true if the features within `self` are a subset of the features within `other`.
+	pub(crate) fn is_subset(&self, other: &Self) -> bool {
+		for (idx, byte) in self.flags.iter().enumerate() {
```
Comment (arik-so) [inline lightning/src/ln/features.rs:678]: I wonder if there's a slightly simpler way to check, bitwise, if `(self & other) == self`. But if there's not, the explicit iteration is extremely legible.

File: lightning/src/ln/monitor_tests.rs [hunk 10]
```diff
@@ -1666,3 +1678,136 @@ fn test_revoked_counterparty_aggregated_claims() {
 	assert!(nodes[1].chain_monitor.chain_monitor.get_and_clear_pending_events().is_empty());
 	assert!(nodes[1].chain_monitor.chain_monitor.get_monitor(funding_outpoint).unwrap().get_claimable_balances().is_empty());
 }
+
+#[cfg(anchors)]
+#[test]
+fn test_yield_anchors_events() {
```
Comment (arik-so) [inline lightning/src/ln/monitor_tests.rs:1684]: mind adding a short blurb about what scenario the test is creating?

File: lightning/src/ln/channel.rs [hunk 11]
```diff
@@ -898,7 +912,18 @@ impl<Signer: Sign> Channel<Signer> {
 			// We've exhausted our options
 			return Err(());
 		}
-		self.channel_type = ChannelTypeFeatures::only_static_remote_key(); // We only currently support two types
+		// We support opening a few different types of channels. Try removing our additional
+		// features one by one until we've either arrived at our default or the counterparty has
+		// accepted one.
+		if self.channel_type.supports_anchors_zero_fee_htlc_tx() {
```
Comment (ariard) [inline lightning/src/ln/channel.rs:918]: Do we have test coverage for `handle_error` once we got an error message due to rejection of channel type ?

**Review by ariard — COMMENTED** [review]
Otherwise test coverage sounds good.

File: lightning/src/ln/monitor_tests.rs [hunk 12]
```diff
@@ -1666,3 +1678,141 @@ fn test_revoked_counterparty_aggregated_claims() {
...
+			(commitment_tx, anchor_tx)
+		},
+		_ => panic!("Unexpected event"),
+	};
+
+	mine_transactions(&nodes[0], &[&commitment_tx, &anchor_tx]);
+	check_added_monitors!(nodes[0], 1);
+
+	let mut holder_events = nodes[0].chain_monitor.chain_monitor.get_and_clear_pending_events();
+	// Certain block ConnectStyle's cause an extra ChannelClose event to be emitted since the best
```
Comment (arik-so) [inline lightning/src/ln/monitor_tests.rs:1751]: nit: `ConnectStyle`s

File: lightning/src/ln/channel.rs [hunk 13]
```diff
@@ -2130,7 +2153,11 @@ impl<Signer: Sign> Channel<Signer> {
 		} else if their_features.supports_channel_type() {
```
Comment (arik-so) [inline lightning/src/ln/channel.rs:2153]: why not just `else if !their_features.supports_channel_type() {`, removing the empty block?

**Review by arik-so — DISMISSED** [review]

**Review by arik-so — APPROVED** [review]
